src/coreComponents/constitutive/solid/DruckerPragerExtended.hpp

```diff
@@ -0,0 +1,495 @@
+                        real64 const x,
+                        real64 & y,
+                        real64 & dy_dx ) const
+  {
+    if( x<0 )
+    {
+      y = y1;
+      dy_dx = 0;
+    }
+    else
+    {
+      y = y1 + (y2-y1)*x/(m+x);
+      dy_dx = (y2-y1)*m/((m+x)*(m+x));
+    }
+  };
+
+};
+
+
+GEOSX_HOST_DEVICE
+GEOSX_FORCE_INLINE
+void DruckerPragerExtendedUpdates::smallStrainUpdate( localIndex const k,
+                                                      localIndex const q,
+                                                      real64 const ( &strainIncrement )[6],
+                                                      real64 ( & stress )[6],
+                                                      real64 ( & stiffness )[6][6] ) const
+{
+  // elastic predictor (assume strainIncrement is all elastic)
+
+  ElasticIsotropicUpdates::smallStrainUpdate( k, q, strainIncrement, stress, stiffness );
+
+  // decompose into mean (P) and von mises (Q) stress invariants
+
+  real64 trialP;
+  real64 trialQ;
+  real64 deviator[6];
+
+  twoInvariant::stressDecomposition( stress,
+                                     trialP,
+                                     trialQ,
+                                     deviator );
+
+  // check yield function F <= 0, using old state
+
+  real64 friction, dfriction_dstate;
+
+  hyperbolicModel( m_initialFriction[k],
+                   m_residualFriction[k],
+                   m_hardening[k],
+                   m_oldState[k][q],
+                   friction,
+                   dfriction_dstate );
+
+  real64 yield = trialQ + friction * (trialP - m_pressureIntercept[k]);
+
+  if( yield < 1e-9 ) // elasticity
+  {
+    return;
+  }
+
+  // else, plasticity (trial stress point lies outside yield surface)
+  // the return mapping can in general be written as a newton iteration.
+
+  real64 solution[3], residual[3], delta[3];
+  real64 jacobian[3][3] = {{}}, jacobianInv[3][3] = {{}};
+
+  solution[0] = trialP; // initial guess for newP
+  solution[1] = trialQ; // initial guess for newQ
+  solution[2] = 1e-5;   // initial guess for plastic multiplier
+
+  real64 norm, normZero = 1e30;
+
+  // begin newton loop
+
+  for( localIndex iter=0; iter<20; ++iter )
+  {
+    m_newState[k][q] = m_oldState[k][q] + solution[2];
+
+    hyperbolicModel( m_initialFriction[k],
+                     m_residualFriction[k],
+                     m_hardening[k],
+                     m_newState[k][q],
+                     friction,
+                     dfriction_dstate );
+
+    // assemble residual system
+    // resid1 = P - trialP + dlambda*bulkMod*dG/dP = 0
+    // resid2 = Q - trialQ + dlambda*3*shearMod*dG/dQ = 0
+    // resid3 = F = 0
+
+    residual[0] = solution[0] - trialP + solution[2] * m_bulkModulus[k] * m_dilationRatio[k] * friction;
+    residual[1] = solution[1] - trialQ + solution[2] * 3 * m_shearModulus[k];
+    residual[2] = solution[1] + friction * (solution[0] - m_pressureIntercept[k]);
+
```

Comment:
May be it will be better for the convergence if we normalize these equation to a dimensionless form. For example we can normalize them to something in order of trialP, trialQ in this case. For other yield function such as DelfEgg or CamClay the third function can be normalized to a value in order of trialP  * trialP or trialQ*trialQ or combination.